### DISCUSSION TITLE
How can I control the DOM programmatically from Quarto?

### DISCUSSION THREAD

**Discussion by AlbertoFabbri93** [quarto-dev/quarto-cli#9780] [Q&A] [answered]
### Description

I wrote the following code to create a Revealjs presentation programmatically from Quarto using R but, even though it is very similar to other people's code I have seen online, it does not work as expected. The two images should be placed side-by-side as specified by the class, instead they are in one column, one above the other:
````qmd
---
title: "A study"
author: "Author"
format: 
  revealjs:
    navigation-mode: grid
    slide-number: h.v
slide-level: 3
execute:
  echo: false
---

```{r}
#| echo: false
#| output: asis

patients <- 1:4

for (patient_num in patients) {
  
  cat(paste("\n## Patient", patient_num))
  
  cat('\n::: {layout-ncol=2}\n')
  
  cat(paste0("![KRT17](Images/Patient_", patient_num, "_plots/Patient_", patient_num, "_krt17.png)\n"))
  
  cat(paste0("![panCK](Images/Patient_", patient_num, "_plots/Patient_", patient_num, "_panCK.png)\n"))
  
  cat(':::\n')

}  
```
````

Can anyone tell me what I am doing wrong?

**Comment by cderv** [comment]
Can you format your code part correctly ? https://quarto.org/bug-reports.html#formatting-make-githubs-markdown-work-for-us 
Thanks ! 

Also, you should look at `keep-md: true` output and see if the generated Markdown is what is expected.  I mean, does the Markdown you are generating is working in revealjs to achieve what you want ? 

This will allow to know if this is something related to how Knitr produce the output or just something not working and not involving the R computation part.

**Reply by AlbertoFabbri93** [reply]
Sorry about the formatting.

I added the `keep-md: true` option and this is the Markdown file I get:

```md
---
title: "A study"
author: "Author"
format: 
  revealjs:
    navigation-mode: grid
    slide-number: h.v
slide-level: 3
execute:
  echo: false
  keep-md: true
---



## Patient 1
::: {layout-ncol=2}
![KRT17](Images/Patient_1_plots/Patient_1_krt17.png)
![panCK](Images/Patient_1_plots/Patient_1_panCK.png)
:::

## Patient 2
::: {layout-ncol=2}
![KRT17](Images/Patient_2_plots/Patient_2_krt17.png)
![panCK](Images/Patient_2_plots/Patient_2_panCK.png)
:::

## Patient 3
::: {layout-ncol=2}
![KRT17](Images/Patient_3_plots/Patient_3_krt17.png)
![panCK](Images/Patient_3_plots/Patient_3_panCK.png)
:::

## Patient 4
::: {layout-ncol=2}
![KRT17](Images/Patient_4_plots/Patient_4_krt17.png)
![panCK](Images/Patient_4_plots/Patient_4_panCK.png)
:::
```

which seems fine to me.

I have a document which generates a very similar Revealjs Markdown file and it works fine (but in that case I wrote the Markdown syntax manually).

**Reply by cderv** [reply]
> The two images should be placed side-by-side as specified by the class, instead they are in one column, one above the other:

I can't reproduce using R cells, or the intermediate markdown shared. 
![Arc_fe21wsPVFz](https://github.com/quarto-dev/quarto-cli/assets/6791940/84bc041a-eede-4f9b-a01e-51d0ad419a9a)


Try with 1.5 pre-release version, we may have fixed this

**Reply by AlbertoFabbri93** [reply]
I tried with Quarto 1.5.39 and I get the same.

```
quarto check
Quarto 1.5.39
[✓] Checking versions of quarto binary dependencies...
      Pandoc version 3.2.0: OK
      Dart Sass version 1.70.0: OK
      Deno version 1.41.0: OK
      Typst version 0.11.0: OK
[✓] Checking versions of quarto dependencies......OK
[✓] Checking Quarto installation......OK
      Version: 1.5.39
      Path: /Users/albi/Applications/quarto/bin

[✓] Checking tools....................OK
      TinyTeX: (not installed)
      Chromium: (not installed)

[✓] Checking LaTeX....................OK
      Tex:  (not detected)

[✓] Checking basic markdown render....OK

[✓] Checking Python 3 installation....OK
      Version: 3.9.6
      Path: /Library/Developer/CommandLineTools/usr/bin/python3
      Jupyter: (None)

      Jupyter is not available in this Python installation.
      Install with python3 -m pip install jupyter

(/) Checking R installation...........'getOption("repos")' replaces Bioconductor standard repositories, see
'help("repositories", package = "BiocManager")' for details.
Replacement repositories:
    BioCsoft: https://bioconductor.org/packages/3.19/bioc
    BioCann: https://bioconductor.org/packages/3.19/data/annotation
    BioCexp: https://bioconductor.org/packages/3.19/data/experiment
    BioCworkflows: https://bioconductor.org/packages/3.19/workflows
    BioCbooks: https://bioconductor.org/packages/3.19/books
    CRAN: https://packagemanager.posit.co/cran/latest
[✓] Checking R installation...........OK
      Version: 4.4.0
      Path: /Library/Frameworks/R.framework/Resources
      LibPaths:
        - /Users/albi/Documents/University/Research_project/brain-metastases-spatial-omics/renv/library/macos/R-4.4/aarch64-apple-darwin20
        - /Users/albi/Library/Caches/org.R-project.R/R/renv/sandbox/macos/R-4.4/aarch64-apple-darwin20/f7156815
      knitr: 1.46
      rmarkdown: 2.27

[✓] Checking Knitr engine render......OK
```

**Reply by cderv** [reply]
> I tried with Quarto 1.5.39 and I get the same.

Get the same as me ? Or get the same problem ? 

I am also rendering with this version. Is what I shared above what you expect or is this unexpected to you ?

**Reply by AlbertoFabbri93** [reply]
> Get the same as me ? Or get the same problem ?

Sorry for not being very clear, I meant that I get the same problem.

> I am also rendering with this version. Is what I shared above what you expect or is this unexpected to you ?

What you shared is what I expect but not what I get, see the following video:


https://github.com/quarto-dev/quarto-cli/assets/51910894/b6b6b1de-43d7-47c2-bf19-13d2007375f4

As you can see, in my presentation the images are aligned vertically instead if horizontally.

**Reply by AlbertoFabbri93** [reply]
> I can't reproduce using R cells, or the intermediate markdown shared. ![Arc_fe21wsPVFz](https://private-user-images.githubusercontent.com/6791940/334332901-84bc041a-eede-4f9b-a01e-51d0ad419a9a.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY5ODQ5OTUsIm5iZiI6MTcxNjk4NDY5NSwicGF0aCI6Ii82NzkxOTQwLzMzNDMzMjkwMS04NGJjMDQxYS1lZWRlLTRmOWItYTAxZS01MWQwYWQ0MTlhOWEuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MjlUMTIxMTM1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjQ0YzA5MTI4NjhlYzAyYTc2YTg4NzY0MjI3MTQyYTAxOGM2YWExNjZhZmUyYTgzZGI3ZjA5ODJhYjhhM2VmMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.k-qhErYw4rqUy59MFdzq88qVmR18n4gmKCUXdeoljaE) 
> 
> Try with 1.5 pre-release version, we may have fixed this

If I use the same 100x100 placeholder images as you I get the same, the images are on the same row (side by side). But I think that that is because the images are small and the horizontal alignment is the default. Can you try with bigger images?

**Reply by mcanouil** [reply]
@AlbertoFabbri93 I suggest you use the provided example below and change it, including the dimensions of the image directly in it until you have a reproducible example to share.

---

You can share a **self-contained "working" (reproducible)** Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).
See <https://quarto.org/bug-reports.html#small-is-beautiful-aim-for-a-single-document-with-10-lines>.

If you have multiple files (and if it is **absolutely required** to have multiple files), please share as a Git repository.

<table>
<tr><th>R</th><th>Python</th></tr>
<tr><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: knitr
---

This is a reproducible Quarto document.

```{r}
x <- c(1, 2, 3, 4, 5)
y <- c(1, 4, 9, 16, 25)

plot(x, y)
```

![An image](https://placehold.co/600x400.png)

The end.
````
`````

</td><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: jupyter
---

This is a reproducible Quarto document.

```{python}
import matplotlib.pyplot as plt

x = [1, 2, 3, 4, 5]
y = [1, 4, 9, 16, 25]

plt.plot(x, y)
plt.show()
```

![An image](https://placehold.co/600x400.png)

The end.
````
`````

</td></tr>
</table>

Additionally and if not already given, please share the output of `quarto check` within a code blocks (*i.e.*, using three backticks ` ```txt `), see <https://quarto.org/bug-reports.html#check>.

**Reply by cderv** [reply, marked as answer]
@AlbertoFabbri93 I missed this but this is because you are not adding images on two paragraphs

````markdown
## Patient 1
::: {layout-ncol=2}
![KRT17](Images/Patient_1_plots/Patient_1_krt17.png)
![panCK](Images/Patient_1_plots/Patient_1_panCK.png)
:::
````

This is only one paragraph in Markdown, so there is nothing to but in the right column. 

You need to add each one in its own paragraph

````markdown
## Patient 1
::: {layout-ncol=2}
![KRT17](Images/Patient_1_plots/Patient_1_krt17.png)

![panCK](Images/Patient_1_plots/Patient_1_panCK.png)
:::
````

So add a newline 
````markdown
---
title: "A study"
author: "Author"
format: 
  revealjs:
    navigation-mode: grid
    slide-number: h.v
slide-level: 3
execute:
  echo: false
keep-md: true
---

```{r}
#| echo: false
#| output: asis

patients <- 1:4

for (patient_num in patients) {
  
  cat(paste("\n## Patient", patient_num))
  
  cat('\n::: {layout-ncol=2}\n')
  
  cat(paste0("![KRT17]({{< placeholder 300 >}})\n"))
  cat("\n")
  cat(paste0("![panCK]({{< placeholder 300 >}})\n"))
  
  cat(':::\n')

}  
```
````

**Reply by AlbertoFabbri93** [reply]
Thank you very much!!
